src/test/java/io/pravega/connectors/flink/FlinkPravegaWriterTest.java

```diff
@@ -546,6 +559,85 @@ public void testTransactionalWriterRestoreState() throws Exception {
+
+            // verify the behavior by passing multiple transaction instance
+            Map<Transaction.Status, List<Transaction<Integer>>> txnMap = new HashMap<>();
+            List<FlinkPravegaWriter.PendingTransaction> pendingTransactionList = new ArrayList<>();
+            int count = 5;
+
+            try (StreamSinkOperatorTestHarness<Integer> testHarness = createTestHarness(context.txnSinkFunction)) {
+                testHarness.setup();
+
+                for (int i = 0; i < count; i++) {
+                    Transaction<Integer> integerTransaction = pendingTransactions.get(i);
+                    Transaction.Status status;
+
+                    if (i % 2 == 0) {
+                        status = Transaction.Status.OPEN;
+                    } else {
+                        if (i == 1) {
+                            status = Transaction.Status.COMMITTING;
+                        } else {
+                            status = Transaction.Status.ABORTED;
+                        }
+                    }
+
+                    switch (i) {
+                        case 0:
+                            pendingTransactionList.add(new FlinkPravegaWriter.PendingTransaction(integerTransaction.getTxnId(), MOCK_SCOPE_NAME_1, MOCK_STREAM_NAME_1));
+                            break;
+                        case 1:
+                            pendingTransactionList.add(new FlinkPravegaWriter.PendingTransaction(integerTransaction.getTxnId(), MOCK_SCOPE_NAME_1, MOCK_STREAM_NAME_2));
+                            break;
+                        case 2:
+                            pendingTransactionList.add(new FlinkPravegaWriter.PendingTransaction(integerTransaction.getTxnId(), MOCK_SCOPE_NAME_1, MOCK_STREAM_NAME_3));
+                            break;
+                        case 3:
+                            pendingTransactionList.add(new FlinkPravegaWriter.PendingTransaction(integerTransaction.getTxnId(), MOCK_SCOPE_NAME_2, MOCK_STREAM_NAME_4));
+                            break;
+                        case 4:
+                            pendingTransactionList.add(new FlinkPravegaWriter.PendingTransaction(integerTransaction.getTxnId(), MOCK_SCOPE_NAME_2, MOCK_STREAM_NAME_5));
+                            break;
+                        default:
+                            break;
+                    }
+
+                    when(integerTransaction.checkStatus()).thenReturn(status);
+
+                    if (txnMap.containsKey(status)) {
+                        txnMap.get(status).add(integerTransaction);
+                    } else {
+                        List<Transaction<Integer>> txnList = new ArrayList<>();
+                        txnList.add(integerTransaction);
+                        txnMap.put(status, txnList);
+                    }
+                }
+                context.txnSinkFunction.restoreState(pendingTransactionList);
+
+                for (Transaction.Status status: txnMap.keySet()) {
```

Comment:
Let me validate my understanding. I think this is checking that the open txn has been invoked commit upon, and for the txns in a different state, commit was not called during restored. Is it what this is validating? If so, could you explain why it is correct that the restore procedure commits the open txn? The other options are that it is left open of it is aborted, and I'd assume that restore may also implement those cases.

Reply:
The restore step takes care of committing all the open transactions that were part of checkpoint but have not been committed yet (due to job failure before the notify checkpoint operation is complete). This test case is modeled to validate various transactions that were part of checkpointed state but has not been committed yet.